### PR TITLE
Refactor unified resources view

### PR DESCRIPTION
**Pull request by gzdunek**
Supports https://github.com/gravitational/teleport/issues/30422

The aim of this PR is extract a common part of the unified resources that can be used both in Web UI and Connect. That shared component should not depend on any Web UI specific APIs, like `useTeleport` or `replaceHistory`.

This PR consists of two commits: the first one extracts a common `Resources` component (the commit is quite large, but it was difficult to separate it into smaller parts), while the second one moves it to the `shared` package. I recommend reviewing them separately.

In some places the `shared/UnifiedResources` component (and subcomponents) imports stuff from the `teleport` package. It is not ideal, but I wanted to limit the number of changes, I will come back to it later.

Key changes:
* The `SharedUnifiedResource` type has been introduced. Its is a union of minimal representations of the resource types. We can't just use types from `teleport` for both Web UI and Connect, because their resource types are not the same (Web UI resources usually contain more properties). Instead, we have minimal types with only the required properties.
* `UnifiedResources` no longer requires `clusterId`. It was used to reset the state of the component when the current cluster has changed. We can achieve the same effect by passing it as a `key`. It will cause the component remount, with no imperative cleanup logic. `clusterId` is also needed for the fetch requests, but it can be supplied at the higher level.
* `guessedAppIconName` has been removed from `App` type, its value was calculated from the other `App` properties, so there is no reason in storing in the object, it can be generated dynamically. 
* `SearchPanel` has been removed from the shared component - the search will look differently in Connect.
* `ResourceCard` is now a clearly presentational component, the responsibility of creating proper cards for each resource type has been moved to `makeUnifiedResourceCard*` functions. While it wasn't necessary, I feel that it is better to have such factory functions, instead of multiple functions conditionally generating each value separately in the child component.
* `replaceHistory` is now called only in the parent component when the search params change, I'm not sure why it was passed to the child components.

This PR should not change any Unified Resources view functionality, let me know if you find any regressions.

### CODE REVIEW

File: web/packages/teleport/src/UnifiedResources/FilterPanel.tsx [hunk 1]
```diff
@@ -23,41 +23,39 @@ import Menu, { MenuItem } from 'design/Menu';
...
+import { SharedUnifiedResource } from './types';
+
+const kindToLabel: Record<SharedUnifiedResource['resource']['kind'], string> = {
+  app: 'Application',
+  db: 'Database',
+  windows_desktop: 'Desktop',
+  kube_cluster: 'Kubernetes',
+  node: 'Server',
+  user_group: 'User group',
+  windows_desktop_service: 'Windows desktop service',
```
Comment (avatus) [inline web/packages/teleport/src/UnifiedResources/FilterPanel.tsx:38]: `windows_desktop_service` isn't used as a kind in the unified cache, we can remove it.

File: web/packages/teleport/src/UnifiedResources/ResourceCard.tsx [hunk 2]
```diff
@@ -552,7 +456,7 @@ const LabelsContainer = styled(Box)`
   overflow: hidden;
 `;
 
-const ResourceLabel = styled(Label)`
+const StyledLabel = styled(Label)`
```
Comment (avatus) [inline web/packages/teleport/src/UnifiedResources/ResourceCard.tsx:462]: Why the name change? I'm ok with it, just curious

File: web/packages/teleport/src/UnifiedResources/Resources.story.tsx [hunk 3]
```diff
@@ -59,49 +54,69 @@ const allResources = [
...
-    <MemoryRouter>
-      <ContextProvider ctx={ctx}>
-        <Resources />
-      </ContextProvider>
-    </MemoryRouter>
+].map(resource => ({
+  resource,
+  ui: {
+    ActionButton: <ButtonBorder size="small">Connect</ButtonBorder>,
+  },
```
Comment (avatus) [inline web/packages/teleport/src/UnifiedResources/Resources.story.tsx]: This is so helpful for access requests too, thank you

File: web/packages/teleport/src/UnifiedResources/cards.ts [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+export function makeUnifiedResourceCardApp(
+  resource: UnifiedResourceApp,
+  ui: UnifiedResourceUi
+): UnifiedResourceCard {
+  return {
+    name: resource.name,
+    SecondaryIcon: ApplicationIcon,
+    primaryIconName: guessAppIcon(resource),
```
Comment (avatus) [inline web/packages/teleport/src/UnifiedResources/cards.ts]: i do lke this being here instead. Good choice

File: web/packages/teleport/src/UnifiedResources/cards.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  return {
+    name: resource.name,
+    SecondaryIcon: ServerIcon,
+    primaryIconName: 'Windows',
+    ActionButton: ui.ActionButton,
+    labels: resource.labels,
+    description: {},
+  };
+}
+
```
Comment (avatus) [inline web/packages/teleport/src/UnifiedResources/cards.ts]: We don't use this. At least, the backend doesn't return these things.

File: web/packages/teleport/src/UnifiedResources/types.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+export type UnifiedResourceUserGroup = {
+  kind: 'user_group';
+  name: string;
+  description: string;
+  labels: ResourceLabel[];
+};
+
+export type UnifiedResourceUi = {
+  ActionButton: React.JSX.Element;
+};
```
Comment (avatus) [inline web/packages/teleport/src/UnifiedResources/types.ts]: amazing 🙌

**Review by avatus — COMMENTED** [review]
This looks great so far. Tested locally with none of my upcoming changes and it works great so far. Going to try a few things with some of my changes on top to make sure we got all that needed to be refactored. Awesome work!!

File: web/packages/shared/components/UnifiedResources/ResourceCard.story.tsx [hunk 7]
```diff
@@ -86,35 +92,51 @@ const Grid = styled.div`
...
+          ),
+        ].map((res, i) => (
+          <ResourceCard
+            key={i}
+            pinned={false}
+            pinResource={() => {}}
+            pinningDisabled={false}
+            selectResource={() => {}}
+            selected={false}
+            pinningNotSupported={false}
```
Comment (avatus) [inline web/packages/shared/components/UnifiedResources/ResourceCard.story.tsx:130]: We will need a third `pinning` related option along the lines of `pinningHidden` or something like that. Might need to rethink the phrase. The idea is that we won't be able to pin (or have the pinned tab) in access requests. Right now, the access request component looks a lot like this story component where we just pass in empty/false. I'm not sure it needs it's own level of abstraction but something to chew on

File: web/packages/shared/components/UnifiedResources/UnifiedResources.story.tsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  aLotOfLabels,
+  ...databases,
+  ...kubes,
+  ...desktops,
+  ...nodes,
+  ...apps,
+  ...databases,
+  ...kubes,
+  ...desktops,
+  ...nodes,
```
Comment (avatus) [inline web/packages/shared/components/UnifiedResources/UnifiedResources.story.tsx:56]: why duplicate spreads?

File: web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+            signal
+          );
+
+          return {
+            startKey: resp.startKey,
+            agents: resp.agents.map(resource => ({
+              resource,
+              ui: {
+                ActionButton: <ResourceActionButton resource={resource} />,
+              },
```
Comment (avatus) [inline web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx:155]: I'm not sure the action button should be returned in the fetch func. Perhaps I'm missing a step but this button wont be 'static' in access requests and requires state from access requests (if its added to the request or not).

File: web/packages/teleport/src/UnifiedResources/FilterPanel.tsx (hunk 1 above)
Comment (gzdunek) [reply web/packages/teleport/src/UnifiedResources/FilterPanel.tsx:38]: In that case, I can remove it completely from the `SharedUnifiedResource` type. 
The only problem is that I won't be able to assign [teleport's `UnifiedResource`](https://github.com/gravitational/teleport/blob/618bb5a81ed90f585d6ab6c1976641e3dd530588/web/packages/teleport/src/services/agents/types.ts#L33) to the shared type, so probably I should remove it from there as well. 
So if I'm correct, that type was wrong from the beginning?

File: web/packages/shared/components/UnifiedResources/ResourceCard.story.tsx (hunk 7 above)
Comment (gzdunek) [reply web/packages/shared/components/UnifiedResources/ResourceCard.story.tsx]: I think we should convert these `pinning*` props to enum:
```ts
export enum PinningSupport {
  Supported = 'Supported',
  /**
   * Disables pinning functionality if a leaf cluster hasn't been upgraded yet.
   * Shows an appropriate message on hover.
   * */
  NotSupported = 'NotSupported',
  /** Disables the pinning button. */
  Disabled = 'Disabled',
  /** Hides the pinning button completely. */
  Hidden = 'Hidden',
}
```

I also added support for hiding.

File: web/packages/shared/components/UnifiedResources/UnifiedResources.story.tsx (hunk 8 above)
Comment (gzdunek) [reply web/packages/shared/components/UnifiedResources/UnifiedResources.story.tsx:56]: I think it was done to show more cards in the story (I didn't change it).

File: web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx (hunk 9 above)
Comment (gzdunek) [reply web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx]: This makes the API a bit more complex, but I think you are right. Fetching and mapping items to the rendered form should be separated.

File: web/packages/shared/components/UnifiedResources/UnifiedResources.tsx [hunk 10]
```diff
@@ -388,9 +385,21 @@ export function UnifiedResources(props: UnifiedResourcesProps) {
   );
 }
 
-export function generateResourceKey({
-  resource,
-}: SharedUnifiedResource): string {
+function getResourcePinningSupport(
+  pinningNotSupported: boolean,
+  updatePinnedResourcesAttempt: Attempt<void>
+): PinningSupport {
```
Comment (avatus) [inline web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:400]: This never returns the new enum `PinningSupport.Hidden`. Perhaps we add it as an argument after `pinningNotSupported` ?

File: web/packages/teleport/src/UnifiedResources/FilterPanel.tsx (hunk 1 above)
Comment (avatus) [reply web/packages/teleport/src/UnifiedResources/FilterPanel.tsx:38]: Yeah seems like it was wrong from the beginning!

File: web/packages/shared/components/UnifiedResources/UnifiedResources.story.tsx (hunk 8 above)
Comment (avatus) [reply web/packages/shared/components/UnifiedResources/UnifiedResources.story.tsx:56]: Ok sounds fine to me

**Review by avatus — APPROVED** [review]
Besides the `Pinning.Hidden` thing all this looks great. I was able to use it to make the `NewAccessRequest` page and things seem to work as usual!

File: web/packages/teleport/src/UnifiedResources/FilterPanel.tsx (hunk 1 above)
Comment (gzdunek) [reply web/packages/teleport/src/UnifiedResources/FilterPanel.tsx]: Thanks, removed!
I also had to fix it in `e`, it uses that incorrect type https://github.com/gravitational/teleport.e/pull/2492.

File: web/packages/shared/components/UnifiedResources/UnifiedResources.tsx (hunk 10 above)
Comment (gzdunek) [reply web/packages/shared/components/UnifiedResources/UnifiedResources.tsx]: Good point, added.

However, while adding it, I realized that we can improve the props passed to `UnifiedResources` a bit.
If we don't support pinning, or just want to hide it, there is no point in passing function to get and update them.
Also, when pinning is not supported, we shouldn't even try to fetch them :/

I refactored all these props to an object that requires get/update functions only when pinning is supported.

Comment (avatus) [reply web/packages/shared/components/UnifiedResources/UnifiedResources.tsx]: Oh yeah, much better

**Review by avatus — APPROVED** [review]

**Comment by gzdunek** [comment]
I pushed the one last fix, when pinning is hidden, we should also hide the 'All resources' and 'Pinned Resources' tabs. Correct, @avatus?

**Comment by avatus** [comment]
> I pushed the one last fix, when pinning is hidden, we should also hide the 'All resources' and 'Pinned Resources' tabs. Correct, @avatus? 

Yes. Sorry I forgot to point that out. I had it that way on my local branch and got confused. Thank you

**Comment by gzdunek** [comment]
Friendly ping @rudream :)

**Review by rudream — APPROVED** [review]

**Pull request merged**